src/lib/operators/join_index.cpp

```diff
@@ -202,40 +208,85 @@ std::shared_ptr<const Table> JoinIndex::_on_execute() {
+      const auto chunk_count_index_input_table = _index_input_table->chunk_count();
+      // We need an order-preserving container here to check whether a chunk of the probe table is not indexed. For
+      // this, we assume (1) the probe table's chunk ids and (2) all indexed chunk ids to be sorted. Starting with
+      // the first elements of these sequences, we successively iterate through the elements and compare the current
+      // values.
+      auto total_indexed_chunk_ids = std::set<ChunkID>{};
+
+      // We do not take multiple table indexes created for the same column into account. For now, we only use the first
+      // available table index. Theoretically, multiple table indexes storing index entries for different chunk subsets,
+      // e.g., ChunkIDs{1,3} and ChunkIDs{0,2,7}, can be used in combination.
+      if (table_indexes.size() > 1) {
+        PerformanceWarning("There are multiple table indexes available, but only the first one is used.");
+      }
 
-      const auto& indexes =
-          index_chunk->get_indexes(std::vector<ColumnID>{_adjusted_primary_predicate.column_ids.second});
+      const auto& table_index = table_indexes.front();
+      const auto& indexed_chunk_ids = table_index->get_indexed_chunk_ids();
+      total_indexed_chunk_ids.insert(indexed_chunk_ids.begin(), indexed_chunk_ids.end());
 
-      if (!indexes.empty()) {
-        // We assume the first index to be efficient for our join
-        // as we do not want to spend time on evaluating the best index inside of this join loop
-        const auto& index = indexes.front();
+      _scan_probe_side_input([&](auto probe_iter, const auto probe_end, const auto probe_chunk_id) {
+        _data_join_probe_segment_with_indexed_segments(probe_iter, probe_end, probe_chunk_id, table_index);
+      });
 
-        // Scan all chunks from the probe side input
-        const auto chunk_count_probe_input_table = _probe_input_table->chunk_count();
-        for (ChunkID probe_chunk_id{0}; probe_chunk_id < chunk_count_probe_input_table; ++probe_chunk_id) {
-          const auto chunk = _probe_input_table->get_chunk(probe_chunk_id);
-          Assert(chunk, "Physically deleted chunk should not reach this point, see get_chunk / #1686.");
+      index_joining_duration += timer.lap();
+
+      // If the chunk was indexed in the table index, there is no need to join it again. Otherwise, perform a
+      // NestedLoopJoin on the not-indexed chunk.
+      auto total_indexed_iter = total_indexed_chunk_ids.begin();
+      for (auto index_side_chunk_id = ChunkID{0}; index_side_chunk_id < chunk_count_index_input_table;
+           ++index_side_chunk_id) {
+        if (*total_indexed_iter == index_side_chunk_id && total_indexed_iter != total_indexed_chunk_ids.end()) {
+          ++total_indexed_iter;
+          ++join_index_performance_data.chunks_scanned_with_index;
+        } else {
```

Comment:
So we fallback if a chunk is not indexed by a table index even though it might be indexed by a chunk index, right? Should not matter if we remove chunk indexes anyway

Reply:
Yes, this is right, table and chunk index based joins are not mixed together.